Avoid dangling-else warnings

We've observed that Bluesim generated code for system tasks can sometimes look like this:
```
  if (!(PORT_RST_N == (tUInt8)0u))
    if (DEF_NOT_cfg_verbosity_read_EQ_0___d9)
      DEF_v__h1782 = dollar_stime(sim_hdl);
    else
      DEF_v__h1782 = 2863311530u;
```
and that GCC will give a warning about the ambiguity of the `else` to human readers:
```
mkMod.cxx: In member function ‘void MOD_mkMod::RL_r()’:
mkMod.cxx:618:6: warning: suggest explicit braces to avoid ambiguous ‘else’ [-Wdangling-else]
   if (!(PORT_RST_N == (tUInt8)0u))
      ^
```
The C++ language specifies that the `else` goes with the nearest `if`, so in the above code, the meaning is what BSC intended.  So I assumed that this warning was intended for human-written code, not machine-generated code, and therefore we didn't need to worry about (but probably want to silence warnings when we can).

However, I'm now wondering if BSC _could_ generate wrong C++ code.

BSC's code-generation for Bluesim is mostly done on an intermediate syntax called `SimCCBlocks`.  These are created in `simMakeCBlocks`, optimized in `simCOpt`, and then written to C++ code in `simBlocksToC`.  The process for writing C++ is to convert the `SimCCBlocks` to an intermediate syntax that closely matches C++ and then pretty-printing that syntax.  This intermediate syntax for C++ is defined in `CCSyntax` and is mostly captured by the `CCFragment` type:
```
data CCFragment =
                ...
                | CBlock [CCFragment]
                ...
                | CIf CCExpr CCFragment (Maybe CCFragment)
                ...
```
As you can see, there is a constructor `CIf` whose arms are `CCFragment`.  The pretty-printer for `CIf` will only insert braces if the arms are `CBlock`.  That is, it's up the creator of the `CCFragment` to specify whether braces should be there or not (based on whether it puts `CBlock` or not).

At the moment, no users of the `CCSyntax` module directly create `CIf` values.  The `CCSyntax` module provides an abstracted function `if_cond`, that takes the condition and the arms, and this function decides what code to actually generate.  (For example, if the condition is a constant, it can just generate the appropriate arm.  And if the arm is an empty block, it can just generate a `CNop` -- although it only does this in the "then" branch?  It does not currently insert braces.)

If `if_cond` is called to create an "if" with an "else" whose "then" clause is an "if" with no "else" ... no braces will be inserted.  I believe that would lead to the "else" being interpreted as going with the wrong "if", right?  BSC only creates if-statements in limited situations, so It's possible that this situation does not occur, so far?

I can see three possible fixes:
1. Make sure that all creators of `CCFragment` values are including `CBlock` as needed
2. Have a cleanup stage, just before pretty printing, that inserts `CBlock` where needed
3. Have the pretty printer insert braces

Options 2 and 3 would seem the least error prone, because they guarantee that the ambiguity won't re-appear if changes are made in the codegen.  However, the situation isn't so bad for 1: Since all creation is done through `if_cond`, it would just need to be addressed in that one place.

Enforcing in the pretty-printer (option 3) isn't too bad, though.  One simple way would be for the pretty-printer to always insert braces.  But this would insert braces in situations that aren't ambiguous, and I assume we want to avoid making the code larger than necessary?

If we want to insert the braces only when necessary, the situation is still not bad.  The pretty-printer uses `printClauseOrBlock` to print the arms of `CIf`, and this function could be passed some info about the context, to be used in deciding how to print the arms.  Specifically: (a) if the context is an `if` with an `else` and the function finds an `if` with no `else`, then it can insert braces; or (b) if the context has no `else` but the branch does, then insert braces.

That seems simple enough, however there is one complication: `printClauseOrBlock` is also used for `CFor` statements.  I assume (but will need to test) that GCC would also warn about dangling-else on code that has an intervening `for` loop:
```
if (...)
  for (...)
    if (...)
      ...
    else
      ...
```
This situation could be addressed by having `printClauseOrBlock` strip away layers of `CFor` when looking for inner `if` statements.  (Which hopefully isn't getting to be too much additional work in the pretty-printer?  Or do we want to just always insert braces around a `for` loop inside an `if` statement?)

At the moment, you can see an example of this by cloning the Flute repo and doing this:
```
cd src_Core/Near_Mem_IO/Near_Mem_IO_AXI4_Unit_Test
make FLUTE_REPO=../../..
```

--- a/src/comp/CCSyntax.hs
+++ b/src/comp/CCSyntax.hs
@@ -675,10 +675,24 @@ instance PPrint CCFragment where
   pPrint d p (CPPInclude file True)  = text ("#include <" ++ file ++ ">")
   pPrint d p (CPPInclude file False) = text ("#include \"" ++ file ++ "\"")
   pPrint d p (CIf c th Nothing)      =
-    (text "if (") <> (pp c) <> (text ")") $+$ (printClauseOrBlock th)
+    -- If the true-arm is a nested if-stmt with an else-clause,
+    -- possibly under layers of for-stmts with single elements,
+    -- then braces are needed to avoid an ambiguous-else warning
+    let needsBraces (CIf _ _ (Just _)) = True
+        needsBraces (CFor _ _ _ b) = needsBraces b
+        needsBraces _ = False
+        th' = if (needsBraces th) then CBlock [th] else th
+    in  (text "if (") <> (pp c) <> (text ")") $+$ (printClauseOrBlock th')
   pPrint d p (CIf c th (Just el))    =
-    (text "if (") <> (pp c) <> (text ")") $+$
-    (printClauseOrBlock th) $+$ (text "else") $+$ (printClauseOrBlock el)
+    -- If the true-arm is a nested if-stmt without an else-clause,
+    -- possibly under layers of for-stmts with single elements,
+    -- then braces are needed for correct parsing
+    let needsBraces (CIf _ _ Nothing) = True
+        needsBraces (CFor _ _ _ b) = needsBraces b
+        needsBraces _ = False
+        th' = if (needsBraces th) then CBlock [th] else th
+    in  (text "if (") <> (pp c) <> (text ")") $+$
+        (printClauseOrBlock th') $+$ (text "else") $+$ (printClauseOrBlock el)
   pPrint d p (CSwitch idx arms deflt) =
     let ppArm (n, blk) = (text "case") <+> (pp n) <> (text ":") $+$
                          nest 2 (vsep (map printStmt blk))

--- a/testsuite/bsc.bluesim/misc/NestedIfBraces.bsv
+++ b/testsuite/bsc.bluesim/misc/NestedIfBraces.bsv
@@ -1,0 +1,28 @@
+
+
+ActionValue #(Bit #(32)) cur_cycle = actionvalue
+                                        Bit #(32) t <- $stime;
+                                        if (genVerilog)
+                                          t = t + 5;
+                                        return t / 10;
+                                     endactionvalue;
+
+(* synthesize *)
+module sysNestedIfBraces ();
+
+  Reg #(Bit #(4)) cfg_verbosity <- mkReg (1);
+
+  Reg #(Bool) rg_state <- mkReg (False);
+
+  rule rl_display (! rg_state);
+    if (cfg_verbosity != 0)
+      $display ("%0d: display", cur_cycle);
+    rg_state <= True;
+  endrule
+
+  rule rl_finish (rg_state);
+    $finish (0);
+  endrule
+
+endmodule
+

--- a/testsuite/bsc.bluesim/misc/misc.exp
+++ b/testsuite/bsc.bluesim/misc/misc.exp
@@ -90,3 +90,13 @@ compile_object_pass AVMethBDPIWithReset.bsv
 copy c_func.c.keep c_func.c
 link_objects_pass {} sysAVMethBDPIWithReset {c_func.c}
 }
+
+# GitHub #442: Nested if-else needs braces when possible ambiguity
+# This tests "if (!RST) if (cond) e1 else e2" in the code for
+# executing system tasks.
+if {$ctest == 1} {
+    compile_object_pass NestedIfBraces.bsv
+    link_objects_pass {} sysNestedIfBraces
+    # Test that there were no warnings during C++ compilation
+    compare_file [make_bsc_ccomp_output_name sysNestedIfBraces]
+}

--- a/testsuite/bsc.bluesim/misc/sysNestedIfBraces.bsc-ccomp-out.expected
+++ b/testsuite/bsc.bluesim/misc/sysNestedIfBraces.bsc-ccomp-out.expected
@@ -1,0 +1,4 @@
+Bluesim object created: sysNestedIfBraces.{h,o}
+Bluesim object created: model_sysNestedIfBraces.{h,o}
+Simulation shared library created: sysNestedIfBraces.cexe.so
+Simulation executable created: sysNestedIfBraces.cexe

